AK+LookupServer: Migrate from DeprecatedFile to File

BenWiederhake opened the pull request:
This PR:
- Switches LookupServer from DeprecatedFile to File
- Implements move-assign and move-construct for HashMap in order to do so efficiently and with few (if any) allocations.

Advances #17129.

Hendiadyoin1 commented on Tests/AK/TestHashMap.cpp:
```diff
@@ -251,3 +251,41 @@ TEST_CASE(take)
     EXPECT(!baz.has_value());
 }
+
+TEST_CASE(move_construct)
+{
+    HashMap<int, int> orig;
+    orig.set(1, 10);
+    orig.set(2, 20);
+    orig.set(3, 30);
+    EXPECT_EQ(orig.size(), static_cast<size_t>(3));
+    EXPECT_EQ(orig.get(2), Optional<int>(20));
+
+    HashMap<int, int> second = move(orig);
+
+    EXPECT_EQ(orig.size(), static_cast<size_t>(0));
+    EXPECT_EQ(orig.get(2), Optional<int>());
+    EXPECT_EQ(second.size(), static_cast<size_t>(3));
+    EXPECT_EQ(second.get(2), Optional<int>(20));
+}
+
+TEST_CASE(move_assign)
+{
+    HashMap<int, int> orig;
+    HashMap<int, int> second;
+    orig.set(1, 10);
+    orig.set(2, 20);
+    orig.set(3, 30);
+
+    EXPECT_EQ(orig.size(), static_cast<size_t>(3));
+    EXPECT_EQ(orig.get(2), Optional<int>(20));
+    EXPECT_EQ(second.size(), static_cast<size_t>(0));
+    EXPECT_EQ(second.get(2), Optional<int>());
+
+    second = move(orig);
+
+    EXPECT_EQ(orig.size(), static_cast<size_t>(0));
+    EXPECT_EQ(orig.get(2), Optional<int>());
+    EXPECT_EQ(second.size(), static_cast<size_t>(3));
+    EXPECT_EQ(second.get(2), Optional<int>(20));
+}
```
To make sure these are truly moves, you can use an `AllocationGuard`, which will assert this on on-target tests

BenWiederhake replied:
Nope, that can't work, since the move ends up calling `HashTable::~HashTable()`, which in turn calls `free`, which does `VERIFY(s_allocation_enabled)`.

Good suggestion, but I don't see how to make it work. Also, it's not really critical anyway.

LucasChollet commented on Userland/Services/LookupServer/LookupServer.cpp:
```diff
@@ -77,24 +79,36 @@ LookupServer::LookupServer()
+        return file_or_error.release_error();
     }
+    auto file = TRY(Core::InputBufferedFile::create(file_or_error.release_value()));
+    auto buffer = TRY(ByteBuffer::create_uninitialized(1024));
```
Can we use `1 * KiB` here?

LucasChollet commented on Userland/Services/LookupServer/LookupServer.cpp:
```diff
@@ -77,24 +79,36 @@ LookupServer::LookupServer()
     u32 line_number = 0;
-    while (!file->eof()) {
-        auto original_line = file->read_line(1024);
+    while (!file->is_eof()) {
```
`while(file->can_read_line())` maybe?

BenWiederhake commented:
Changes since last push:
- Changed `1024` to `1 * KiB`
- Changed `while (!file->eof()) {` to `while (TRY(file->can_read_line())) {`, thanks to @LucasChollet

AtkinsSJ commented on Userland/Services/LookupServer/LookupServer.cpp:
```diff
@@ -77,24 +79,36 @@ LookupServer::LookupServer()
+    auto new_hosts_or_error = this->try_load_etc_hosts();
+    if (new_hosts_or_error.is_error())
+        dbgln("Ignoring '/etc/hosts', keeping old values");
+    else {
```
Nit: braces on all branches if they're on one.

BenWiederhake replied:
Background: I was experimenting with other things and inserted a few `dbgln` temporarily in the else branch, and apparently forgot to remove the braces. Whoops!

AtkinsSJ commented on Userland/Services/LookupServer/LookupServer.cpp:
```diff
@@ -77,24 +79,36 @@ LookupServer::LookupServer()
+{
+    HashMap<Name, Vector<Answer>, Name::Traits> map;
+    auto add_answer = [&map](Name const& name, RecordType record_type, DeprecatedString data) {
+        map.ensure(name).empend(name, record_type, RecordClass::IN, s_static_ttl, move(data), false);
```
`try_ensure` and `try_empend` since we're already propagating errors here.

AtkinsSJ commented on Userland/Services/LookupServer/LookupServer.cpp:
```diff
@@ -122,6 +136,8 @@ void LookupServer::load_etc_hosts()
         builder.append(".in-addr.arpa"sv);
```
`try_append`, and on line above. (I can't seem to select multiple lines using my phone. 😅 )

BenWiederhake commented:
Changes since last push:
- Rebase for convenience
- Fix bad braces (thanks @AtkinsSJ!)
- Use `try_{ensure,empend,append}`, given that we're in an error-handling context anyway.

BenWiederhake commented:
Changes since last push:
- Rebase for convenience
- Dropped the changes to AK, as they were effectively just adding a FIXME, which is more appropriate in #18821.

ADKaster commented on Userland/Services/LookupServer/LookupServer.cpp:
```diff
@@ -77,24 +79,35 @@ LookupServer::LookupServer()
+{
+    HashMap<Name, Vector<Answer>, Name::Traits> map;
+    auto add_answer = [&map](Name const& name, RecordType record_type, DeprecatedString data) {
+        return map.ensure(name).try_empend(name, record_type, RecordClass::IN, s_static_ttl, move(data), false);
```
The resolution to <https://github.com/SerenityOS/serenity/pull/18759#discussion_r1192111603> seems to have been lost. This should use try_ensure.

BenWiederhake replied:
Thanks! It wasn't lost, I just focused too much on the `empend` call, and somehow overlooked the `ensure` call in the same line. Will fix!

BenWiederhake commented:
Changes since last push:
- Turned an overlooked `map.ensure` into a `map.try_ensure` (Thanks @ADKaster!)
- Rebased for convenience.

However, since `HashMap::try_ensure` returns `ErrorOr<V>` instead of `ErrorOr<V&>`, this causes an unnecessary copy of the answer vector (which should be very short anyway, but causes quadratic runtime nevertheless), and requires another lookup to get the ensured value. The core of the problem seems to be that `ErrorOr<V&>` is impossible because `Variant` only accepts `NotLvalueReference`. Fixing this is far outside the scope of this PR.
